pkg/report: get only the thread local reports

tarasmadan opened the pull request:
Syzkaller is currently able to parse sequential reports from the kernel log and accumulate the statistics per report. You can see these reports in syzkaller only (not syzbot dashboard).
The problems:
1. These reports may be interleaving if are coming from the parallel contexts (threads)
2. Getting reports from one thread instead of all threads will hopefully give a higher signal/noise ratio.

This PR improves sequential reports parsing and is not expected to affect "first report" extraction.
Eventually the plan is to disable panic_on_warn and start getting sequential reports.

TODO:
1. Consider using MAX_REPORT_LINES.
1. Consider skipping next report if previous report truncated to the next one becomes invalid.
1. Comment code?

ramosian-glider reviewed (commented):
I think this code should also handle reports from the CPU context.

tarasmadan commented on vm/vm.go:
```diff
@@ -493,7 +493,19 @@ func (mon *monitor) extractErrors(defaultError string) []*report.Report {
+func getReportChain(reps []*report.Report, threadID string) []*report.Report {
+	var res []*report.Report
+	for _, rep := range reps {
+		if rep.ThreadID == threadID {
```
> Haven't pkg/report already filtered reports by context?
No. It extracts the reports one-by-one paying attention to the "current report ThreadID".

> If no, then how have pkg/report parsed output with intermixed reports from different context? Are we careful enough during parsing to properly split all intermixed reports from different contexts? Or if not, then isn't our parsing broken?

Current logic is trying to get report assuming it may be starting at any line.
Suppressed/broken reports are ignored.

dvyukov replied:
> No. It extracts the reports one-by-one paying attention to the "current report ThreadID".

Then isn't it broken for what you want to do?
What will it extract from the following log?
```
[C1] report0_first_line
[C1] report0_middle_line
[C1] report0_last_line
[C2] report1_first_line
[C2] report1_middle_line
[C1] report2_first_line
[C1] report2_middle_line
[C1] report2_last_line
[C2] report1_last_line
```

tarasmadan replied:
I think it will work correctly because report1 will be parsed from `report1_first_line` to `report1_last_line` ignoring the lines from another context.
Once done, we'll start looking for the next report from the line `[C2] report1_middle_line`.
It is a bit ~quadratish algorithm, but the correctness should be Ok.

Let me add the synthetic test to cover this case. It should be helpful to double check your concern and prevent these problems in the future.

tarasmadan replied:
> Will anything here affect emails?

After this PR and #6332 the plan is to disable panic_on_warn.
We'll start receiving bloated first report after that. That is the problem.
Thanks. It means point 1 is critical and should be fixed.

dvyukov commented on pkg/report/report_test.go:
```diff
@@ -545,3 +552,26 @@ func TestSplitReportBytes(t *testing.T) {
+// TestParseAll's focus points are the ability to:
+// 1. parse multiple reports
+// 2. extract correct ThreadID/CpuID.
+func TestParseAll(t *testing.T) {
```
> Do you feel concerned only about the parse_all or also about the guilty, guilty_raw, symbolize etc.?

guilty accepts different inputs (symbolized and cleaned output, basically reports).
I dunno what is symbolize, and why it's separated.

dvyukov replied:
One file per test was intentional, it makes diff reviews possible.
For example, if we split them into 2 files, if somebody changes something that changes output, PR will contain only the output files, but it will be hard to understand if these are legit changes or not, b/c you won't see the input files in the PR.

tarasmadan committed:
pkg/report: get only the thread local reports

Syzkaller is currently able to parse sequential reports from the kernel log and accumulate the statistics per report.
You can see these reports in syzkaller only (not syzbot dashboard).
The problems:
1. These reports may be interleaving if are coming from the parallel contexts (threads)
2. Getting reports from one thread instead of all threads will hopefully give a higher signal/noise ratio.

This PR improves sequential reports parsing and is not expected to affect "first report" extraction.

tarasmadan committed:
pkg/report/testdata/linux: add testdata for ContextIDs and tail reports

And regenerate test with tail reports and ContextIDs.

dvyukov commented on pkg/report/testdata/starnix/report/3:
```diff
@@ -236,3 +236,153 @@ internal error: entered unreachable code: Symlink nodes cannot be opened.
 [00173.546403][2244][3058][klog] INFO: 0x00004272309b3b18: ff330738 000041df 309b3c20 00004272 |8.3..A.. <.0rB..|
 [00173.546406][2244][3058][klog] INFO: 0x00004272309b3b28: cd6abe2d 000043c5 6a69aac8 0000426b |-.j..C....ijkB..|
+TAIL REPORTS:
+2024/06/12 21:38:34 SYZFATAL: failed to recv *flatrpc.HostMessageRaw: EOF
```
same here

dvyukov reviewed (commented):
Please check other test changes for sanity.

tarasmadan commented on pkg/report/testdata/linux/report/626:
```diff
@@ -56,3 +56,33 @@ Local variable ----path@step_into created at:
  path_openat+0x3603/0x64c0
 =====================================================
+TAIL REPORTS:
+ prepare_task_switch+0x284/0xd00
```
Title line is "Kernel panic - not syncing: panic_on_kmsan set ..."
Because it is a report with Thread IDs, we save up to 50 prefix lines.
The parsing was started after the "BUG: KMSAN: uninit-value in prepare_task_switch+0x284/0xd00" thus all the lines (~18) were included.

We can alternatively remove all the prefix lines for the tail reports.
